docs/adrs/005-gh-branch-strategy.md

```diff
@@ -0,0 +1,22 @@
+# 5. Branching Strategy in Github
+
+Date: 2023-03-29
+
+## Status
+
+Proposed
+
+## Context
+
+RKE2 is released at the same cadence as upstream Kubernetes. This requires management of multiple versions at any given point in time. The current branching strategy uses `release-v[MAJOR].[MINOR]`, with the `master` branch corresponding to the highest version released based on [semver](https://semver.org/). Github's Tags are then used to cut releases, which are just point-in-time snapshots of the specified branch at a given point. As there is the potential for bugs and regressions to be on present on any given branch, this branching and release strategy requires a code freeze to QA the branch without new potentially breaking changes going in.
+
+## Decision
+
+We will introduce additional branches as `dev-v[MAJOR].[MINOR]` that will be where all standard development will occur. The `release-v[MAJOR].[MINOR]` branches will still be used to release from.
+
+## Consequences
+
+- Allows for constant development and no code freeze.
+- Any critical CVEs that would necessitate releasing again outside of the standard cadence can be added directly to the `release` branches and then ported back to the `dev` branches.
```

Comment:
Is this to avoid additional tagging? The code freeze would still exist on the release branches, but not the dev branches, right? So in the case of a critical CVE which requires an additional release, the release team should cherry-pick from the dev branch to get the fix. Then, after release, we rebase to fast forward the release branch to the dev branch? It kinda seems like we could achieve the same results just using tags, right? (it would rearrange history, but so does the branch solution)

Reply:
It seems like what you want to get closer to is trunk development with release trains, but we need to isolate changes using a feature flagging system to make that possible.
One simple way to implement this kind of isolation without rearranging our entire system is to commit commented out code. Then create a PR to uncomment the code after release. There are ways to make this easier (open/close principle is one), and other ways to implement the isolation, but this is the easiest that comes to mind.